### DISCUSSION TITLE
Group by Record Description in Invoice

### DISCUSSION THREAD

**Discussion by iDontWantAUsername** [kimai/kimai#5041] [General]
Hi,
I was just wondering if it is possible to group by the timesheet record description in the generated invoice? I have had a look at the twig file but i am not sure it is possible to manipulate the data in here.

My use case is that i log my time on various projects and activities but include a description for each of those for the specific 'thing' i am working on, however during the same day i might work in that thing before and after lunch. Currently these will appear as 2 lines in the Invoice when grouped by Day, ideally these would be combined into one line. e.g.:
```
09:00->11:00 - Activity 1 - description of work
11:00->13:00 - Activity 2 - some other description
14:00->18:00 - Activity 1 - description of work
```
This appears in the invoice as
```
01/02/2024 
description of work
some other description
description of work
```

Ideally this would be able to group by the same description string and just appear as:
```
01/02/2024 
description of work
some other description
```

In https://github.com/kimai/invoice-templates/blob/638b6419411a7abc4c66a5397cad8cda3628821a/din5008-invoice/din5008-invoice.pdf.twig#L390C1-L392C40 i think it is these lines in the twig that display the description:

```
                            {% if entry.description is not empty %}
                                <br><small>{{ entry.description }}</small>
                            {% endif %}
```
I think i would need to group prior to this, i assume this would be code as opposed to an template? If possible in an invoice template that would be great.

Thanks in advance

**Comment by kevinpapst** [comment]
If you group by day, all line items are merged and the description kept as you described it.
But that ONLY works if all entries have the same hourly rate.

This is the code that does the merging:
https://github.com/kimai/kimai/blob/main/src/Invoice/Calculator/DateInvoiceCalculator.php

This groups by date and user, you could create one that does the same for date and description for example:
https://github.com/kimai/kimai/blob/main/src/Invoice/Calculator/DateUserInvoiceCalculator.php

**Reply by iDontWantAUsername** [reply]
Thanks, i tried adding the additional group to the DateInvoiceCalculator.php:
```
        return [
            $invoiceItem->getBegin()->format('Y-m-d'),
            $invoiceItem->getDescription()
        ];
```

Unfortunately that the means a separate entry line in the invoice for each date and description.

This is what i am trying to achieve:

![image](https://github.com/user-attachments/assets/27bac9f7-8590-4b65-9e0d-6f6a58e37f25)

It might just be too complex and not achievable with the current architecture.
